### PR TITLE
Laboratory work #4, Tasya Kakoulina - 23FPL2

### PR DESCRIPTION
**Pull request by tasya-kakoulina**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -40,6 +45,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        Vectorizer.__init__(self, self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:49]: use `super`

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -51,6 +59,10 @@ def set_tokenized_corpus(self, tokenized_corpus: TokenizedCorpus) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(tokenized_corpus, list) or not tokenized_corpus:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:63]: add message to the error

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -67,6 +79,12 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list) or not tokenized_document:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:83]: The same

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -67,6 +79,12 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
...
+        vector = self._calculate_bm25(tokenized_document)
+        if not vector:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:86]: the same

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -81,6 +99,16 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document or not isinstance(tokenized_document, list):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:103]: the samw

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -81,6 +99,16 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+        vector_list = [0.0 * n for n in range(len(self._vocabulary))]
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:104]: `0.0 * n` -> `0.0` I believe you wanted another thing

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -81,6 +99,16 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+        k1 = 1.5
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:105]: do not create these variables

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -113,6 +145,17 @@ def put_corpus(self, corpus: Corpus) -> None:
                 or if input arguments are empty,
                 or if methods used return None.
         """
+        if not corpus:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:149]: the same

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -113,6 +145,17 @@ def put_corpus(self, corpus: Corpus) -> None:
...
+        self.__documents = corpus
+        tokenized_docs = [tok_doc for doc in self.__documents if (tok_doc := self._tokenizer.tokenize(doc))]
+        if not tokenized_docs:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:153]: the same

File: lab_4_retrieval_w_clustering/main.py [hunk 10]
```diff
@@ -170,6 +229,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
+        BasicSearchEngine.__init__(self, self._db.get_vectorizer(), self._db.get_tokenizer())
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:233]: `super`

File: lab_4_retrieval_w_clustering/main.py [hunk 11]
```diff
@@ -182,6 +243,16 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if not isinstance(query, str) or not query or not isinstance(n_neighbours, int) or n_neighbours <= 0:
+            raise ValueError
+        tokenized_query = self._db.get_tokenizer().tokenize(query)
+        vectorized_query = self._db.get_vectorizer().vectorize(tokenized_query)
+        vectors = [tup[1] for tup in self._db.get_vectors()]
+        neighbours = self._calculate_knn(vectorized_query, vectors, n_neighbours)
+        if not neighbours:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:253]: the same

File: lab_4_retrieval_w_clustering/main.py [hunk 12]
```diff
@@ -227,11 +302,15 @@ def set_new_centroid(self, new_centroid: Vector) -> None:
             ValueError: In case of inappropriate type input arguments,
                 or if input arguments are empty.
         """
+        if not isinstance(new_centroid, tuple) or not new_centroid:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:306]: the same

File: lab_4_retrieval_w_clustering/main.py [hunk 13]
```diff
@@ -244,6 +323,10 @@ def add_document_index(self, index: int) -> None:
             ValueError: In case of inappropriate type input arguments,
                 or if input arguments are empty.
         """
+        if not isinstance(index, int) or index is None or index < 0:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:327]: the same

File: lab_4_retrieval_w_clustering/settings.json [hunk 14]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 8
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/settings.json:2]: What about 10?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:105]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:104]: but it works fine - it creates a list with n number of zeros

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:233]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:49]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: is it better?..

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (artyomtugaryov) [reply lab_4_retrieval_w_clustering/main.py:114]: Of course, remove `* n`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nice, I like it

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 12 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 13 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done?..

File: lab_4_retrieval_w_clustering/main.py [hunk 15]
```diff
@@ -81,6 +109,16 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list):
+            raise ValueError('Invalid document input')
+        vector_list = [0.0] * len(self._vocabulary)
+        bm_dict = calculate_bm25(self._vocabulary, tokenized_document, self._idf_values,
+                                 1.5, 0.75, self._avg_doc_len, len(tokenized_document))
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Do not set `1.5, 0.75` is default value

File: lab_4_retrieval_w_clustering/main.py [hunk 16]
```diff
@@ -170,6 +243,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
+        super().__init__(self._db.get_vectorizer(), self._db.get_tokenizer())
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: should be the first line in this method

File: lab_4_retrieval_w_clustering/start.py [hunk 17]
```diff
@@ -49,7 +59,27 @@ def main() -> None:
...
+    documents = ''.join(open_files()[0])
+    paragraphs = get_paragraphs(documents)
+    stopwords = open_files()[-1]
+    query = 'Первый был не кто иной, как Михаил Александрович Берлиоз, председатель правления'
+    db = DocumentVectorDB(stopwords)
+    db.put_corpus(paragraphs)
+    vector_search = VectorDBSearchEngine(db)
+    result = vector_search.retrieve_relevant_documents(query, 3)
+    print(result)
+    # for num in range(1, 15):
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/start.py]: what is it?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 15 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 16 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/start.py (hunk 17 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/start.py]: changed...

File: lab_4_retrieval_w_clustering/main.py [hunk 18]
```diff
@@ -3,9 +3,21 @@
...
 Vector search with clusterization
 """
+from json import dump
+from math import sqrt
+
+from lab_2_retrieval_w_bm25.main import calculate_bm25
 
 # pylint: disable=undefined-variable, too-few-public-methods, unused-argument, duplicate-code, unused-private-member, super-init-not-called
-from lab_3_ann_retriever.main import BasicSearchEngine, Tokenizer, Vector, Vectorizer
+from lab_3_ann_retriever.main import (
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:10]: All imports should be together

File: lab_4_retrieval_w_clustering/main.py [hunk 19]
```diff
@@ -40,6 +55,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        super().__init__(self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: should be the first line in the init

File: lab_4_retrieval_w_clustering/main.py [hunk 20]
```diff
@@ -140,6 +199,13 @@ def get_vectors(self, indices: list[int] | None = None) -> list[tuple[int, Vecto
         Returns:
             list[tuple[int, Vector]]: List of index and vector for documents.
         """
+        if not indices:
+            return list(self.__vectors.items())
+        unique = []
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:202]: set?

File: lab_4_retrieval_w_clustering/main.py [hunk 21]
```diff
@@ -154,6 +220,13 @@ def get_raw_documents(self, indices: tuple[int, ...] | None = None) -> Corpus:
         Returns:
             Corpus: List of documents.
         """
+        if not indices:
+            return self.__documents
+        unique = []
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:223]: set?

File: lab_4_retrieval_w_clustering/main.py [hunk 22]
```diff
@@ -287,6 +398,28 @@ def run_single_train_iteration(self) -> list[ClusterDTO]:
...
+                distances.append((distance, centroids.index(centroid)))
+            closest = min(distances)
+            self.__clusters[closest[-1]].add_document_index(vectors.index(vector))
+        for cluster in self.__clusters:
+            cluster_vectors = []
+            for num in cluster.get_indices():
+                cluster_vectors.append(vectors[num][-1])
+            new_centroid = tuple(sum(value) / len(value) for value in zip(*cluster_vectors))
+            cluster.set_new_centroid(new_centroid)
+        return self.__clusters
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Looks like the method should return new clusters, but not set it to the `self`

File: lab_4_retrieval_w_clustering/main.py [hunk 23]
```diff
@@ -427,8 +654,8 @@ def retrieve_relevant_documents(
 
         Returns:
             list[tuple[float, str]] | None: Relevant documents with their distances.
-
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:430]: return

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 19 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 18 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:10]: i'm not really sure what you mean, hope it's better now but i honestly let isort decide how to import

File: lab_4_retrieval_w_clustering/main.py (hunk 20 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:202]: no cause i can't change the order of elements inside indices

File: lab_4_retrieval_w_clustering/main.py (hunk 21 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:223]: no cause i can't change the order of elements inside indices

File: lab_4_retrieval_w_clustering/main.py (hunk 23 above)
Comment (tasya-kakoulina) [reply lab_4_retrieval_w_clustering/main.py:430]: done

**Pull request closed without merging**